Sources/ChatToys/Utils/WebContext.swift

```diff
@@ -194,3 +198,67 @@ extension Array {
     }
 }
 
+public class MarkdownProcessor {
+    
+    public static var stringToKeyMap = [String: String]()
+    public static var uuidToURLMap = [String: String]()
+
+    public static func key(for value: String) -> String? {
+        return stringToKeyMap.first { $0.value == value }?.key
+    }
+
+    public static func url(for value: String) -> URL? {
+        guard let textContent = key(for: value)?.nilIfEmptyOrJustWhitespace,
+              var urlString = uuidToURLMap[value] else {
+            return nil
+        }
+        
+        let words = textContent.split(separator: " ")
+        
+        // TODO see if a more robust implementation would be warranted
+        if let range = urlString.range(of: "#:~:text=") {
+            let urlString = urlString[..<range.lowerBound]
+        }
+        
+        if words.count <= 8 {
+            urlString = "\(urlString)#:~:text=\( processWords(words.prefix(8)) )"
+        } else {
+            let start = processWords(words.prefix(4))
+            let end = processWords(words.suffix(4))
+            urlString = "\(urlString)#:~:text=\(start),\(end)"
+        }
+        
+        return URL(string: urlString)
+    }
+
+    private static func processWords(_ words: ArraySlice<Substring>) -> String {
+        return words.joined(separator: " ")
+            .addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
+            .replacingOccurrences(of: ",", with: "%2C")
+            .replacingOccurrences(of: "*", with: "")
+            .replacingOccurrences(of: "#", with: "")
+            .replacingOccurrences(of: "-", with: "")
+            .replacingOccurrences(of: "_", with: "")
+            .trimmingCharacters(in: .whitespaces)
+    }
+
+    public static func shortUUID(_ string: String) -> String {
+        if let existingKey = stringToKeyMap[string] {
+            return existingKey
+        }
+        var uuid: String
+        repeat {
+            uuid = String(UUID().uuidString.prefix(4))
+        } while stringToKeyMap.values.contains(uuid)
```

Comment:
This will start slowing down as you get close to the max # of uuid 4 character prefixes.  At some point (after assigning a lot of ids) we should probably increase the number of chars. Here's one idea: if you hit a collision twice in a row, increase the prefix length by 1 and try again?